MLflow and Airflow tutorial (#2348)
astronomer/docs: learn/airflow-mlflow.md

---
title: "Use MLflow with Apache Airflow"
sidebar_label: "MLflow"

> description: "How to produce to use the MLflow Airflow provider"

Reviewer: I'm not sure what to suggest here, but this description seems off....guessing it got "half updated" at some point?

id: airflow-mlflow
sidebar_custom_props: { icon: 'img/integrations/mlflow.png' }
---

import CodeBlock from '@theme/CodeBlock';
import mlflow_tutorial_dag from '!!raw-loader!../code-samples/dags/airflow-mlflow/mlflow_tutorial_dag.py';

> [MLflow](https://mlflow.org/) is a commonly used tool for tracking and managing machine learning models. It can be used together with Airflow for ML orchestration (MLOx), leveraging both tools for what they do best. In this tutorial, you’ll learn about three different ways you can use MLflow with Airflow.

Reviewer 2: ```suggestion
[MLflow](https://mlflow.org/) is a popular tool for tracking and managing machine learning models. It can be used together with Airflow for ML orchestration (MLOx), leveraging both tools for what they do best. In this tutorial, you’ll learn about three different ways you can use MLflow with Airflow.
```

:::info

> If you are already familiar with MLflow and Airflow and want to get a use case up and running, check out this [quickstart repository](https://github.com/astronomer/learn-airflow-mlflow-tutorial) which automatically starts up Airflow and local MLflow and MinIO instances. Clone the quickstart repository, configure the local MLflow connection as shown in [Step 2](#step-2-configure-your-airflow-connection) of this tutorial, and run the three example DAGs.

Reviewer 2: ```suggestion
If you're already familiar with MLflow and Airflow and want to get a use case up and running, clone the [quickstart repository](https://github.com/astronomer/learn-airflow-mlflow-tutorial) to automatically start up Airflow and local MLflow and MinIO instances. After you clone the repository, configure the local MLflow connection as shown in [Step 2](#step-2-configure-your-airflow-connection) of this tutorial, and run the three example DAGs.
```

:::

## Three ways to use MLflow with Airflow

> The DAG in this tutorial shows three different ways Airflow can interact with MLFlow:

Reviewer 2: ```suggestion
The DAG in this tutorial shows three different ways Airflow can interact with MLflow:
```

> - Using an MLflow operator from the [MLflow Airflow provider](https://github.com/astronomer/airflow-provider-mlflow): The MLflow provider contains several operators that abstract over common actions you might want to perform in MLflow, for example creating a deployment with the [CreateDeploymentOperator](https://github.com/astronomer/airflow-provider-mlflow/blob/main/mlflow_provider/operators/deployment.py) or running predictions from an existing model with the [ModelLoadAndPredictOperator](https://github.com/astronomer/airflow-provider-mlflow/blob/main/mlflow_provider/operators/pyfunc.py). 

Reviewer 2: ```suggestion
- Use an MLflow operator from the [MLflow Airflow provider](https://github.com/astronomer/airflow-provider-mlflow). The MLflow provider contains several operators that abstract over common actions you might want to perform in MLflow, such as creating a deployment with the [CreateDeploymentOperator](https://github.com/astronomer/airflow-provider-mlflow/blob/main/mlflow_provider/operators/deployment.py) or running predictions from an existing model with the [ModelLoadAndPredictOperator](https://github.com/astronomer/airflow-provider-mlflow/blob/main/mlflow_provider/operators/pyfunc.py). 
```

> - Using an MLflow hook from the MLflow Airflow provider: The MLflow provider contains several [Airflow hooks](what-is-a-hook.md) that allow you to connect to MLflow using credentials stored in an Airflow connection. You can use these hooks if you need to perform actions in MLflow for which no dedicated operator exists. You can also use these hooks to create your own [custom operators](airflow-importing-custom-hooks-operators.md).

Reviewer 2: ```suggestion
- Use an MLflow hook from the MLflow Airflow provider. The MLflow provider contains several [Airflow hooks](what-is-a-hook.md) that allow you to connect to MLflow using credentials stored in an Airflow connection. You can use these hooks if you need to perform actions in MLflow for which no dedicated operator exists. You can also use these hooks to create your own [custom operators](airflow-importing-custom-hooks-operators.md).
```

> - Using an MLflow hook from the MLflow Airflow provider: The MLflow provider contains several [Airflow hooks](what-is-a-hook.md) that allow you to connect to MLflow using credentials stored in an Airflow connection. Using these hooks is ideal to perform actions in MLflow from within an Airflow tasks for which no dedicated operator exists. You can also use these hooks to create your own [custom operators](airflow-importing-custom-hooks-operators.md).

Reviewer: ```suggestion
- Using an MLflow hook from the MLflow Airflow provider: The MLflow provider contains several [Airflow hooks](what-is-a-hook.md) that allow you to connect to MLflow using credentials stored in an Airflow connection. You can use these hooks if you need to perform actions in MLflow for which no dedicated operator exists. You can also use these hooks to create your own [custom operators](airflow-importing-custom-hooks-operators.md).
```

> - Using the MLflow Python package directly in a [@task decorated task](airflow-decorators.md): The MLflow Python package contains functionality like tracking metrics and artifacts with [`mlflow.sklearn.autolog`](https://mlflow.org/docs/latest/python_api/mlflow.sklearn.html). You can use this package to write custom Airflow tasks for ML-related actions like feature engineering.

Reviewer 2: ```suggestion
- Use the MLflow Python package directly in a [@task decorated task](airflow-decorators.md). The MLflow Python package contains functionality like tracking metrics and artifacts with [`mlflow.sklearn.autolog`](https://mlflow.org/docs/latest/python_api/mlflow.sklearn.html). You can use this package to write custom Airflow tasks for ML-related actions like feature engineering.
```

> - Using the MLflow Python package directly in a [@task decorated task](airflow-decorators.md): When performing ML related actions in an Airflow task such as feature engineering you can track metrics and artifacts using the MLflow Python package directly like in any Python script. For example with [`mlflow.sklearn.autolog`](https://mlflow.org/docs/latest/python_api/mlflow.sklearn.html). 

Reviewer: ```suggestion
- Using the MLflow Python package directly in a [@task decorated task](airflow-decorators.md): The MLflow Python package contains functionality like tracking metrics and artifacts with [`mlflow.sklearn.autolog`](https://mlflow.org/docs/latest/python_api/mlflow.sklearn.html). You can use this package to write custom Airflow tasks for ML-related actions like feature engineering.
```

- Airflow operators. See [Operators 101](what-is-an-operator.md).
- Airflow hooks. See [Hooks 101](what-is-a-hook.md).
- Airflow connections. See [Managing your Connections in Apache Airflow](connections.md).

## Prerequisites

- The [Astro CLI](https://docs.astronomer.io/astro/cli/get-started).
- An [MLflow instance](https://www.mlflow.org/docs/latest/quickstart.html). This tutorial uses a local instance.
- An object storage connected to your MLflow instance. This tutorial uses [MinIO](https://min.io/).

## Step 1: Configure your Astro project

1. Create a new Astro project:

    ```sh
    $ mkdir astro-mlflow-tutorial && cd astro-mlflow-tutorial
    $ astro dev init
    ```

2. Add the following packages to your `packages.txt` file:

    ```text
    git
    gcc
    gcc python3-dev
    ```

3. Add the following packages to your `requirements.txt` file:

    ```text
    airflow-provider-mlflow==1.1.0
    mlflow-skinny==2.3.2
    ```

## Step 2: Configure your Airflow connection

To connect Airflow to your MLflow instance, you need to create a [connection in Airflow](connections.md). 

> 1. Run `astro dev start` in your Astro project to start up Airflow and navigate to the Airflow UI at `localhost:8080`.

Reviewer 2: ```suggestion
1. Run `astro dev start` in your Astro project to start up Airflow and open the Airflow UI at `localhost:8080`.
```

2. In the Airflow UI, go to **Admin** -> **Connections** and click **+**.

3. Create a new connection named `mlflow_default` and choose the `HTTP` connection type. Enter the following values to create a connection to a local MLflow instance:

>     - **Connection ID**: `mlflow_default`.
>     - **Connection Type**: `HTTP`.

Reviewer 2: ```suggestion
    - **Connection ID**: `mlflow_default`
    - **Connection Type**: `HTTP`
```

    - **Host**: `http://host.docker.internal`
    - **Port**: `5000`

:::info 

> If you are using a remote MLflow instance, enter your MLflow instance URL as the **Host** and your username and password as the **Login** and **Password** fields in the connection. If you are running your MLflow instance via Databricks, provide your Databricks URL to the **Host** field, enter the string `token` in the **Login** field and your [Databricks personal access token](https://docs.databricks.com/dev-tools/auth.html#personal-access-tokens-for-users) in the **Password** field.

Reviewer 2: ```suggestion
If you are using a remote MLflow instance, enter your MLflow instance URL as the **Host** and your username and password as the **Login** and **Password** in the connection. If you are running your MLflow instance via Databricks, enter your Databricks URL as the **Host**, enter `token` as the **Login** and your [Databricks personal access token](https://docs.databricks.com/dev-tools/auth.html#personal-access-tokens-for-users) as the **Password**.
```

> If you are using a remote MLflow instance, you need to enter the MLflow instance URL as a **Host** and provide your username to the **Login** and your password to the **Password** field. To run your MLflow instance via Databricks, provide your Databricks URL to the **Host** field, enter the string `token` in the **Login** field and your [Databricks personal access token](https://docs.databricks.com/dev-tools/auth.html#personal-access-tokens-for-users) in the **Password** field.

Reviewer: ```suggestion
If you are using a remote MLflow instance, enter your MLflow instance URL as the **Host** and your username and password as the **Login** and **Password** fields in the connection. If you are running your MLflow instance via Databricks, provide your Databricks URL to the **Host** field, enter the string `token` in the **Login** field and your [Databricks personal access token](https://docs.databricks.com/dev-tools/auth.html#personal-access-tokens-for-users) in the **Password** field.
```

Please note that the **Test** button might return a 405 error message even if your credentials are correct. 

:::

## Step 3: Create your DAG

1. In your `dags` folder, create a file called `mlflow_tutorial_dag.py`.

2. Copy the following code into the file. Make sure to provide the name of a bucket in your object storage that is connected to your MLflow instance to the `ARTIFACT_BUCKET` variable.

    <CodeBlock language="python">{mlflow_tutorial_dag}</CodeBlock>

    This DAG consists of three tasks, each showing a different way to use MLflow with Airflow.

    - The `create_experiment` task creates a new experiment in MLflow by using the [MLflowClientHook](https://github.com/astronomer/airflow-provider-mlflow/blob/main/mlflow_provider/hooks/client.py) in a TaskFlow API task. The MLflowClientHook is one of several [hooks](https://github.com/astronomer/airflow-provider-mlflow/tree/main/mlflow_provider/hooks) in the MLflow provider that contains abstractions over calls to the MLflow API. 

>     - The `scale_features` task shows a way to use MLflow and Airflow without the MLflow provider. In this task the [mlflow](https://pypi.org/project/mlflow/) package is used directly with [scikit-learn](https://pypi.org/project/scikit-learn/) to log information about the scaler to MLflow.

Reviewer: ```suggestion
    - The `scale_features` task uses the [mlflow](https://pypi.org/project/mlflow/) package in a Python decorated task with [scikit-learn](https://pypi.org/project/scikit-learn/) to log information about the scaler to MLflow. This functionality is not included in any modules of the MLflow provider, so a custom Python function is the best way to implement this task.
```

>     - The `create_registered_model` task shows how to use the [CreateRegisteredModelOperator](https://github.com/astronomer/airflow-provider-mlflow/blob/main/mlflow_provider/operators/registry.py) to register a new model in your MLflow instance.

Reviewer: ```suggestion
    - The `create_registered_model` uses the [CreateRegisteredModelOperator](https://github.com/astronomer/airflow-provider-mlflow/blob/main/mlflow_provider/operators/registry.py) to register a new model in your MLflow instance.
```

## Step 4: Run your DAG

> 1. Manually run the `mlflow_tutorial_dag` DAG by clicking the play button.

Reviewer 2: ```suggestion
1. In the Airflow UI run the `mlflow_tutorial_dag` DAG by clicking the play button.
```

    ![DAGs overview](/img/guides/mlflow_tutorial_dag_graph_view.png)

2. Open the MLflow UI (if you are running locally at `localhost:5000`) to see the data recorded by each task in your DAG.

    The `create_experiment` task created the `Housing` experiments, where your `Scaler` run from the `scale_features` task was recorded.

    ![MLflow UI experiments](/img/guides/mlflow_experiments.png)

    The `create_registered_model` task created a registered model with two tags.

    ![MLflow UI models](/img/guides/mlflow_registered_models.png)

3. Open your object storage (if you are using a local MinIO instance at `localhost:9000`) to see your MLflow artifacts.

    ![MinIO experiment artifacts](/img/guides/mlflow_experiment_artifacts_in_minio.png)

## Conclusion

> Congratulations! You used MLflow and Airflow together in three different ways. Learn more about other operators and hooks in the MLflow Airflow provider in the [documentation](https://github.com/astronomer/airflow-provider-mlflow).

Reviewer: If we don't publish documentation for MLFlow provider before this tutorial gets published (likely will not...), let's update "documentation" to "the official repo" or something like that. Don't want to over-promise here...
